core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java

```diff
@@ -99,4 +162,50 @@ public ManagedChannelImpl build() {
    */
   @Internal
   protected abstract ClientTransportFactory buildTransportFactory();
+
+  private static class DirectAddressNameResolver implements NameResolver {
+    final SocketAddress address;
+
+    DirectAddressNameResolver(SocketAddress address) {
+      this.address = address;
+    }
+
+    @Override
+    public Handle resolve(String scheme, @Nullable String nsAuthority, String endpoint) {
+      return new Handle() {
+        @Override
+        public String getServiceAuthority() {
+          if (address instanceof InetSocketAddress) {
```

Comment:
This still concerns me, because it seems transport-specific and I feel like it will get forgotten and bite us. Could the constructor become `AbstractManagedChannelImplBuilder(SocketAddress directServerAddress, String authority)` so that the transport still explicitly decides what the default authority name is?